EIPS/eip-4841.md

```diff
@@ -0,0 +1,420 @@
+ */
+function addStorage(address storageAddr_) public virtual returns (uint256) {
+    _assets.push(AssetStorage({
+        addr: storageAddr_,
+        stock: IStorageContract(storageAddr_)
+    }));
+    return _assets.length-1; // index
+}
+```
+
+### Property Layer
+
+A contract layer that stores the attribute values for which SVG tag to use.
+
+The user can access the saved image through the Property Layer, and inheritance of EIP-721 or other standards is also performed in this layer.
+
+The following functions are descriptions of the main functions to be implemented in the Property Layer's contract.
+
+- `getImage(uint256)`: This function gets the saved SVG image. Get the property values corresponding to `tokenId_` and call `getImage` of the Assemble Layer interface.
+
+  ```solidity
+  /**
+  * @dev See {IAssembleContract-getImage}
+  */
+  function getImage(uint256 tokenId_) public view virtual returns (string memory) {
+      return assembleContract.getImage(_attrs[tokenId_]);
+  }
+  ```
+
+- `setAssembleContract(address)`: This function sets the Assemble Layer's Contract. If you want to use it as an Immutable Storage that cannot be changed after the initial deployment, you can remove the function.
+
+  ```solidity
+  /**
+  * @param newAssembleContractAddr_ Address value of `AssembleContract` to be changed.
+  * @dev If later changes or extensions are unnecessary, write directly to `constructor` without implementing the function.
+  */
+  function setAssembleContract(address newAssembleContractAddr_) public virtual {
+      assembleContract = IAssembleContract(newAssembleContractAddr_);
+  }
+  ```
+
+- `_setAttr(uint256)`: This is a function that sets the attribute value of which SVG image tag to load. The logic for how to set the property value **MUST** be implemented separately according to the direction to be used.
+
+  ```solidity
+  /**
+  * @param tokenId_ The token ID for which you want to set the attribute value.
+  * @dev Set the attribute value of the corresponding `tokenId_` sequentially according to the number of asset storage.
+  */
+  function _setAttr(uint256 tokenId_) internal virtual {
+      for (uint256 idx=0; idx < assembleContract.getStorageCount(); idx++) {
+          uint256 newValue = 0;
+
+          /// @dev Implement the property value setting logic.
+
+          _attrs[tokenId_].push(newValue);
+      }
+  }
+  ```
+
+## Rationale
+
+### Large Capacity Storage
+
+The best way for us to keep our content permanent and tamper-proof is to store it on Ethereum, rather than on centralized servers or IPFS, where it can be tampered with or lost. Like the SVG format, various extensions have come out to reduce the size of the content, but most of the content still has a size of several MB or more. Through this EIP, we would like to provide a solution that can safely store SVG images in sizes ranging from tens of KB to several MB on Ethereum.
```

Comment:
```suggestion
The best way for us to keep our content permanent and tamper-proof is to store it on Ethereum, rather than on centralized servers or IPFS, where it can be tampered with or lost. Like the SVG format, various extensions have come out to reduce the size of the content, but most of the content still has a size of several megabytes or more. Through this EIP, we would like to provide a solution that can safely store SVG images in sizes ranging from tens of KB to several MB on Ethereum.
```